### PR TITLE
CASSANDRA-16244 Create a jvm upgrade dtest for mixed versions repairs

### PR DESCRIPTION
**Pull request by adelapena**

### CODE REVIEW

File: test/distributed/org/apache/cassandra/distributed/upgrade/MixedModeRepairTest.java [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+    @Test
+    public void testMixedModeRepair() throws Throwable
+    {
+        new UpgradeTestBase.TestCase()
+        .nodes(2)
+        .nodesToUpgrade(2)
+        .upgrade(Major.v30, Major.v4)
+        .upgrade(Major.v3X, Major.v4)
+        .withConfig(config -> config.with(NETWORK, GOSSIP))
+        .setup(cluster -> cluster.schemaChange(withKeyspace(CREATE_TABLE)))
```
Comment (yifan-c) [inline test/distributed/org/apache/cassandra/distributed/upgrade/MixedModeRepairTest.java]:
```suggestion
        .setup(cluster -> {
            cluster.schemaChange(CREATE_TABLE);
            cluster.setUncaughtExceptionsFilter(throwable -> throwable instanceof RejectedExecutionException);
        })
```

The test fails with `RejectedExecutionException` when shutting down occasionally. We can set the uncaught exception filter to ignore the error. 

The full trace:

```
INFO  [ACCEPT-/127.0.0.1] node1 2021-01-25 15:30:44,734 MessagingService.java:1129 - MessagingService has terminated the accept() thread

org.apache.cassandra.distributed.shared.ShutdownException: Uncaught exceptions were thrown during test

  at org.apache.cassandra.distributed.impl.AbstractCluster.checkAndResetUncaughtExceptions(AbstractCluster.java:862)
  at org.apache.cassandra.distributed.impl.AbstractCluster.close(AbstractCluster.java:848)
  at org.apache.cassandra.distributed.upgrade.UpgradeTestBase$TestCase.run(UpgradeTestBase.java:185)
  at org.apache.cassandra.distributed.upgrade.MixedModeRepairTest.testMixedModeRepair(MixedModeRepairTest.java:106)
  ...
  Suppressed: java.util.concurrent.RejectedExecutionException: ThreadPoolExecutor has shut down
    at org.apache.cassandra.concurrent.DebuggableThreadPoolExecutor$1.rejectedExecution(DebuggableThreadPoolExecutor.java:61)
    ...
    at org.apache.cassandra.concurrent.DebuggableThreadPoolExecutor.execute(DebuggableThreadPoolExecutor.java:151)
    at org.apache.cassandra.net.MessagingService.receive(MessagingService.java:873)
    at org.apache.cassandra.net.IncomingTcpConnection.receiveMessage(IncomingTcpConnection.java:212)
    at org.apache.cassandra.net.IncomingTcpConnection.receiveMessages(IncomingTcpConnection.java:181)
    at org.apache.cassandra.net.IncomingTcpConnection.run(IncomingTcpConnection.java:92)
```

Comment (yifan-c) [reply test/distributed/org/apache/cassandra/distributed/upgrade/MixedModeRepairTest.java]: No need to wrap the `CREATE_TABLE` with `withKeyspace()` again.

**Review by yifan-c — COMMENTED** [review]
The test flow looks good. 

The reason that `repair` issued from the lower versioned node is not finishing is that the `PrepareMessage` timeout is 1 HOUR! So it is unlikely to see the repair failure in the jvm dtest.
After changing the timeout to 1 second in the 3.0 and 3.11 code base respectively, the repair will throw the following error. Since we are unable to update the timeout of the message, the solution in the test sounds valid to me. 

As Alexander Dejanovski pointed out, can we have dedicated check for the each node? 
* Assert the repair should fail with the expected log message in the 4.0 version node. 
* Assert the repair in the 3.0/3.x version node should timeout. 

```
Error:
java.lang.RuntimeException: Repair job has failed with the error message: [2021-01-25 16:02:07,351] Repair command #1 failed with error Got negative replies from endpoints [127.0.0.2]
	at org.apache.cassandra.tools.RepairRunner.progress(RepairRunner.java:115)
	at org.apache.cassandra.utils.progress.jmx.JMXNotificationProgressListener.handleNotification(JMXNotificationProgressListener.java:77)
	at javax.management.NotificationBroadcasterSupport.handleNotification(NotificationBroadcasterSupport.java:275)
	at javax.management.NotificationBroadcasterSupport$SendNotifJob.run(NotificationBroadcasterSupport.java:352)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)



	at org.apache.cassandra.distributed.api.NodeToolResult$Asserts.fail(NodeToolResult.java:214)
	at org.apache.cassandra.distributed.api.NodeToolResult$Asserts.errorContainsAny(NodeToolResult.java:118)
	at org.apache.cassandra.distributed.api.NodeToolResult$Asserts.errorContains(NodeToolResult.java:110)
	at org.apache.cassandra.distributed.upgrade.MixedModeRepairTest.lambda$testMixedModeRepair$3(MixedModeRepairTest.java:91)
	at org.apache.cassandra.distributed.upgrade.UpgradeTestBase$TestCase.run(UpgradeTestBase.java:180)
	at org.apache.cassandra.distributed.upgrade.MixedModeRepairTest.testMixedModeRepair(MixedModeRepairTest.java:113)
```

File: test/distributed/org/apache/cassandra/distributed/upgrade/MixedModeRepairTest.java (hunk 1 above)
Comment (adelapena) [reply test/distributed/org/apache/cassandra/distributed/upgrade/MixedModeRepairTest.java]: Good catches, fixed as suggested.

**Comment by adelapena** [comment]
Thanks for the feedback. Indeed the repair takes ages to fail in the < 4.0 nodes due to that timeout in `PrepareMessage`. I have separated the cases for the repair in the upgraded and in the not upgraded node, where we only don't wait for the repair in the not upgraded node.

**Pull request closed without merging**